src/Storage/AuthorizationCodeStorage.php

```diff
@@ -5,120 +5,90 @@
-		'code'         => 'string', // authorization code.
 		'client_id'    => 'string', // client identifier.
-		'user_login'   => 'string', // The WordPress user id.
 		'redirect_uri' => 'string', // redirect URI.
 		'expires'      => 'int',    // expires as unix timestamp.
 		'scope'        => 'string', // scope as space-separated string.
 		'id_token'     => 'string', // The OpenID Connect id_token.
 	);
 
-	public function __construct() {
-		// Store the authorization codes in a taxonomy.
-		register_taxonomy( self::TAXONOMY, null );
-		foreach ( self::$authorization_code_data as $key => $type ) {
-			register_term_meta(
-				self::TAXONOMY,
-				$key,
-				array(
-					'type'              => $type,
-					'single'            => true,
-					'sanitize_callback' => array( __CLASS__, 'sanitize_' . $key ),
-				)
-			);
+	private function getUserIdByCode( $code ) {
+		if ( empty( $code ) ) {
+			return null;
 		}
-	}
-
-	public static function sanitize_string_length( $string, $length ) {
-		return substr( $string, 0, $length );
-	}
 
-	public static function sanitize_code( $code ) {
-		return self::sanitize_string_length( $code, 40 );
-	}
+		// @akirk: get_users() is better than a direct db query: in a Multiuser installation, it adds a query that checks whether the user is a member of the blog.
+		$users = get_users(
+			array(
+				'number'       => 1,
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_key
+				'meta_key'     => self::META_KEY_PREFIX . '_client_id_' . $code,
+				// Using a meta_key EXISTS query is not slow, see https://github.com/WordPress/WordPress-Coding-Standards/issues/1871.
+				'meta_compare' => 'EXISTS',
+			)
+		);
+
+		if ( empty( $users ) ) {
+			return null;
+		}
 
-	public static function sanitize_client_id( $client_id ) {
-		return self::sanitize_string_length( $client_id, 200 );
+		return absint( $users[0]->ID );
 	}
 
-	public static function sanitize_redirect_uri( $redirect_uri ) {
-		return self::sanitize_string_length( $redirect_uri, 2000 );
-	}
+	public function getAuthorizationCode( $code ) {
+		$user_id = $this->getUserIdByCode( $code );
+		if ( empty( $user_id ) ) {
+			return null;
+		}
 
-	public static function sanitize_scope( $scope ) {
-		return self::sanitize_string_length( $scope, 100 );
-	}
+		$authorization_code = array(
+			'user_id' => $user_id,
+			'code'    => $code,
+		);
 
-	public static function sanitize_id_token( $id_token ) {
-		return self::sanitize_string_length( $id_token, 2000 );
-	}
+		foreach ( array_keys( self::$authorization_code_data ) as $key ) {
+			$authorization_code[ $key ] = get_user_meta( $user_id, self::META_KEY_PREFIX . '_' . $key . '_' . $code, true );
+		}
 
-	public static function sanitize_user_login( $user_login ) {
-		return self::sanitize_string_length( $user_login, 60 );
-	}
+		if ( $authorization_code['expires'] < time() ) {
+			// Remove it right now to avoid cleanup work later.
+			$this->expireAuthorizationCode( $code );
```

Comment:
@akirk This really doesn't avoid cleanup work later, since this would only cleanup when the client was too late to query for token, which in real world would be quite rare. In fact, its not possible to avoid cleanup, so that's best addressed in a cron or an admin page in #34 and I think we should remove it from here to keep things simple.

Reply:
I thought it was useful for my testing but I am indifferent about it.

Reply:
Alright! I will remove and merge the PR.